privacy/bolton/optimizers.py

```diff
@@ -0,0 +1,390 @@
+               optimizer,
+               loss,
+               dtype=tf.float32,
+              ):
+    """Constructor.
+
+    Args:
+      optimizer: Optimizer_v2 or subclass to be used as the optimizer
+        (wrapped).
+      loss: StrongConvexLoss function that the model is being compiled with.
+      dtype: dtype
+    """
+
+    if not isinstance(loss, StrongConvexMixin):
+      raise ValueError('loss function must be a Strongly Convex and therefore '
+                       'extend the StrongConvexMixin.')
+    self._private_attributes = ['_internal_optimizer',
+                                'dtype',
+                                'noise_distribution',
+                                'epsilon',
+                                'loss',
+                                'class_weights',
+                                'input_dim',
+                                'n_samples',
+                                'layers',
+                                'batch_size',
+                                '_is_init'
+                               ]
+    self._internal_optimizer = optimizer
+    self.learning_rate = GammaBetaDecreasingStep()  # use the Bolton Learning
+    # rate scheduler, as required for privacy guarantees. This will still need
+    # to get values from the loss function near the time that .fit is called
+    # on the model (when this optimizer will be called as a context manager)
+    self.dtype = dtype
+    self.loss = loss
+    self._is_init = False
+
+  def get_config(self):
+    """Reroutes to _internal_optimizer. See super/_internal_optimizer."""
+    return self._internal_optimizer.get_config()
+
+  def project_weights_to_r(self, force=False):
+    """Normalize the weights to the R-ball.
+
+    Args:
+      force: True to normalize regardless of previous weight values.
+        False to check if weights > R-ball and only normalize then.
+
+    Raises:
+      Exception:
+    """
+    if not self._is_init:
+      raise Exception('This method must be called from within the optimizer\'s '
+                      'context.')
+    radius = self.loss.radius()
+    for layer in self.layers:
+      weight_norm = tf.norm(layer.kernel, axis=0)
+      if force:
+        layer.kernel = layer.kernel / (weight_norm / radius)
+      else:
+        layer.kernel = tf.cond(
+            tf.reduce_sum(tf.cast(weight_norm > radius, dtype=self.dtype)) > 0,
+            lambda k=layer.kernel, w=weight_norm, r=radius: k / (w / r),  # pylint: disable=cell-var-from-loop
+            lambda k=layer.kernel: k  # pylint: disable=cell-var-from-loop
+        )
+
+  def get_noise(self, input_dim, output_dim):
+    """Sample noise to be added to weights for privacy guarantee.
+
+    Args:
+      input_dim: the input dimensionality for the weights
+      output_dim: the output dimensionality for the weights
+
+    Returns:
+      Noise in shape of layer's weights to be added to the weights.
+
+    Raises:
+      Exception:
```

Comment:
Please detail here. Make sure the line ends with a `.` and has no trailing whitespace.